### ISSUE TITLE
Embed LR schedule and initialization with the model

### ISSUE THREAD

**Issue by Maratyszcza** [pytorch/vision#36]
I tried to implement [SqueezeNet](https://github.com/DeepScale/SqueezeNet) as a `torchvision` model and train it via [ImageNet example](https://github.com/pytorch/examples/tree/master/imagenet), and found that it doesn't converge as is. The reference code differs in two aspect:

- All but the last convolutions are initialized with Xavier Glorot initializer, the last is normal with stdev 0.01
- The learning rate is linearly decreased (polynomial schedule with power=1).

In PyTorch these aspects are hard-coded inside the ImageNet example, but I think it makes sense to make them part of the model definition in `torch.vision`. What's your position on it?

**Comment by apaszke** [comment]
Weights can be initialized in model's `__init__`, so it has nothing to do with the imagenet example, right?

As for lr schedule, I think we can just do sth like
```python
if hasattr(model, 'lr_schedule'):
    lr = model.lr_schedule(epoch)
else:
    lr = args.lr * (0.1 ** (epoch // 30))
```

**Comment by colesbury** [comment]
I've been trying to put weight initialization as part of the model, since it often seem particular to the type of architecture. I added it to the ResNet definition and I'm going to add it to the VGG model def:
https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py#L112

I'm not sure about learning rate schedule. It seems awkward to put it as part of the model definition, but as you point out, hard coding it in the ImageNet example isn't ideal either

**Comment by Maratyszcza** [comment]
@apaszke @colesbury Thanks, I will do the same for SqueezeNet. Are there pre-defined initialization functions/classes for popular initialization schemes (e.g. like in [Neon](http://neon.nervanasys.com/index.html/initializers.html) or [Keras](https://keras.io/initializations/))?

@colesbury I think ideally we should provide a default learning schedule as a part of `torch.vision` models and let users override it via command-line arguments.

**Comment by apaszke** [comment]
Yes, we should add them in nn somewhere.

**Comment by eladhoffer** [comment]
What do you think about this kind of regime inside of the model? https://github.com/eladhoffer/convNet.pytorch/models/alexnet.py

**Comment by apaszke** [comment]
Can't open it, are you sure the link is correct and the repo is public?

**Comment by eladhoffer** [comment]
How about https://raw.githubusercontent.com/eladhoffer/convNet.pytorch/master/models/alexnet.py
The repo is public

**Comment by apaszke** [comment]
That's one way to approach it, but I'm not sure if it's the most convenient one. Having a function that returns an optimizer for a given epoch seems more powerful.

**Comment by alykhantejani** [comment]
Is there/will there be a nice way to adapt the learning rate or momentum but keep other state in the optimizer, i.e. for Adam

**Comment by alykhantejani** [comment]
Should this issue be moved to the `pytorch` repo instead?

**Comment by vfdev-5** [comment]
As you, guys, speak also here about weight initialization, what about DenseNet if we want to use a not pretrained model ? 
According to [Caffe official implementation](https://github.com/liuzhuang13/DenseNetCaffe/blob/master/make_densenet.py), convolutions are initialized with something like `kaiming_normal`.
